projects/sandbox/data_generation/data_generation/bilby_injections.py

```diff
@@ -0,0 +1,177 @@
+):
+
+    configure_logging(logdir / "generate_testing_set.log", verbose)
+
+    datadir = datadir / "bilby"
+    datadir.mkdir(exist_ok=True, parents=True)
+    logdir.mkdir(exist_ok=True, parents=True)
+
+    signal_files_exist = all(
+        [
+            Path(datadir / f"{ifo}_bilby_injections.hdf5").exists()
+            for ifo in ifos
+        ]
+    )
+
+    if signal_files_exist and not force_generation:
+        logging.info(
+            "Bilby timeseries already exists and forced generation is off. "
+            "Not generating bilby timeseries."
+        )
+        return
+
+    # query the first coincident contiguous segment
+    # of required minimum duration and download the data
+    segment_names = [f"{ifo}:{state_flag}" for ifo in ifos]
+
+    # generate signal times based on requested spacing
+    # and save as text file for bilby
+    spacing += waveform_duration // 2
+    signal_times = np.linspace(
+        buffer + waveform_duration // 2, n_samples * spacing, n_samples
+    )
+    # bilby requests the start time of the signals
+    # so subtract half the bilby
+    # duration which will enforce that the signal
+    # lies in the center of the kernel
+    np.savetxt(
+        datadir / "start_times.txt", signal_times - (bilby_duration / 2)
+    )
+    min_duration = (
+        signal_times[-1] - signal_times[0] + waveform_duration // 2 + buffer
+    )
+
+    segment_start, segment_stop = query_segments(
+        segment_names,
+        start,
+        stop,
+        min_duration,
+    )[0]
+
+    background_dict = download_data(
+        ifos, frame_type, channel, sample_rate, segment_start, segment_stop
+    )
+    df = 1 / waveform_duration
+    if gaussian:
+        logging.info(
+            "Generating gaussian noise from psd for injection background"
+        )
+        for ifo in ifos:
+            duration = len(background_dict[ifo]) / sample_rate
+            psd = normalize_psd(background_dict[ifo], df, sample_rate)
+            data = noise_from_psd(psd, df, duration, sample_rate)
+            background_dict[ifo] = TimeSeries(data, dt=1 / sample_rate)
+
+    # instantiate prior, sample, and generate signals
+    prior = prior()
+    parameters = prior.sample(n_samples)
+    signals = generate_gw(
+        parameters,
+        sample_rate,
+        waveform_duration,
+        waveform,
+        waveform_arguments=waveform_arguments,
+        parameter_conversion=parameter_conversion,
+    )
+
+    plus, cross = signals.transpose(1, 0, 2)
+    plus = torch.Tensor(plus)
+    cross = torch.Tensor(cross)
+
+    # project raw polarizations onto interferometers
+    # with sampled sky localizations
+    tensors, vertices = get_ifo_geometry(*ifos)
+
+    # dec is declination
+    # psi is polarization angle
+    # phi is relative azimuthal angle between source and earth
+    dec = torch.Tensor(parameters["dec"])
+    psi = torch.Tensor(parameters["psi"])
+    phi = torch.Tensor(parameters["ra"]) - np.pi
```

Comment:
This is incorrect. What is the coordinate system of get_ifo_geometry

Reply:
If it is the celestial frame, then the reference for the azimuthal coordinate is gmst. You can convert the time of the injection to gmst either using astropy or lal. I checked that both give consistent value. Ex. the start time in our project config is
```
$ lal_tconvert 1262751694
Sat Jan 11 04:21:16 UTC 2020
```

```
>>> lal.GreenwichMeanSiderealTime(lal.LIGOTimeGPS(1262751694)) % (2*np.pi)
3.0625954275359035
>>> t = Time(1262751694, format='gps', scale='utc')
>>> t.sidereal_time('mean', 'greenwich').to('rad')
<Longitude 3.06258219 rad>
```
which are similar up to 4th decimal place